File: shared/src/commonMain/kotlin/com/snabbit/runner/shared/core/camera/CaptureRegistry.kt
package com.snabbit.runner.shared.core.camera

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Optional synchronous mirror of the registry's token→path map, for a consumer
 * that must resolve a path WITHOUT suspending — the Android `WebViewAssetLoader`
 * callback runs on a background worker thread and cannot call the suspend
 * [CaptureRegistry.resolve]. The registry owns the lifecycle: it calls [put] on
 * register and [remove] on eviction/expiry, so there is a single source of truth
 * (the mirror can't drift from the registry). Implementations MUST be thread-safe.
 */
interface CaptureIndex {
    fun put(token: String, filePath: String)
    fun remove(token: String)
}

/**
 * In-memory registry of captured media, keyed by opaque tokens with
 * automatic TTL-based expiry.
 *
 * Ported from Flutter's `capture_registry.dart`. Key differences:
 * - **Thread-safe:** Uses [Mutex] instead of relying on Dart's
 *   single-threaded model.
 * - **Coroutine-based sweep:** Replaces `Timer.periodic` with a
 *   coroutine `delay()` loop scoped to [sweepScope].
 * - **Unguessable tokens:** [generateToken] draws from a per-platform CSPRNG.
 *
 * Captures are registered after the camera writes a file to the
 * dedicated temp directory, and resolved by the asset-loader path
 * handler when the WebView fetches the image/video URL.
 *
 * @param index optional synchronous mirror ([CaptureIndex]) kept in lock-step
 *   with the internal map for the WebView asset-loader's non-suspending lookup.
 */
class CaptureRegistry(
    private val sweepScope: CoroutineScope,
    private val defaultTtlMs: Long = DEFAULT_TTL_MS,
    private val sweepIntervalMs: Long = SWEEP_INTERVAL_MS,
    private val onDeleteFile: suspend (String) -> Unit = {},
    private val index: CaptureIndex? = null,
    private val clock: () -> Long = { currentTimeMillis() },
) {
    private val mutex = Mutex()
    private val entries = mutableMapOf<String, CaptureEntry>()
    private var sweepRunning = false

    /**
     * Registers a file path and returns the opaque token the WebView
     * can use to fetch the media via the asset-loader URL.
     *
     * If [token] is provided it is used as-is; otherwise one is generated
     * from the platform CSPRNG.
     */
    suspend fun register(
        filePath: String,
        token: String? = null,
        ttlMs: Long = defaultTtlMs,
    ): String {
        val key = token ?: generateToken()
        val expiresAt = clock() + ttlMs
        mutex.withLock {
            entries[key] = CaptureEntry(filePath, expiresAt)
            index?.put(key, filePath)
            ensureSweepRunning()
        }
        return key
    }

    /**
     * Returns the file path for [token] if it is still valid, or `null`
     * if the token is unknown or expired. Expired entries are evicted
     * eagerly — the entry is removed under the lock and its backing file
     * deleted *after* the lock is released (see [ensureSweepRunning]).
     */
    suspend fun resolve(token: String): String? {
        val expiredPath = mutex.withLock {
            val entry = entries[token] ?: return null
            if (clock() <= entry.expiresAt) return entry.filePath
            // Expired — evict the entry now, delete its file after unlocking.
            entries.remove(token)
            index?.remove(token)
            entry.filePath
        }
        safeDeleteFile(expiredPath)
        return null
    }

    /**
     * Explicitly evicts [token] — removes the entry and deletes the backing
     * file (outside the lock). No-op if the token is unknown.
     */
    suspend fun evict(token: String) {
        val path = mutex.withLock {
            val entry = entries.remove(token) ?: return
            index?.remove(token)
            entry.filePath
        }
        safeDeleteFile(path)
    }

    /** Returns the number of currently registered entries. */
    suspend fun size(): Int = mutex.withLock { entries.size }

    // ── Private helpers ──────────────────────────────────────────

    private fun ensureSweepRunning() {
        if (sweepRunning) return
        sweepRunning = true
        sweepScope.launch {
            while (true) {
                delay(sweepIntervalMs)
                val (expiredPaths, idle) = mutex.withLock {
                    val now = clock()
                    val expired = entries.filterValues { now > it.expiresAt }
                    expired.keys.forEach { key ->
                        entries.remove(key)
                        index?.remove(key)
                    }
                    val paths = expired.values.map { it.filePath }
                    if (entries.isEmpty()) {
                        sweepRunning = false
                        paths to true
                    } else {
                        paths to false
                    }
                }
                // Delete backing files OUTSIDE the lock: onDeleteFile is suspend I/O,
                // and holding the (non-reentrant) mutex across it would serialize the
                // capture hot path — and deadlock if the callback re-entered the registry.
                expiredPaths.forEach { safeDeleteFile(it) }
                if (idle) break
            }
        }
    }

    private suspend fun safeDeleteFile(filePath: String) {
        try {
            onDeleteFile(filePath)
        } catch (_: Throwable) {
            // Best-effort cleanup — file may already be gone.
        }
    }

    private data class CaptureEntry(
        val filePath: String,
        val expiresAt: Long,
    )

    companion object {
        const val DEFAULT_TTL_MS = 15L * 60 * 1000 // 15 minutes
        const val SWEEP_INTERVAL_MS = 5L * 60 * 1000 // 5 minutes

        /** Photo size limit: 10 MB */
        const val MAX_PHOTO_SIZE_BYTES = 10L * 1024 * 1024

        /** Video size limit: 50 MB */
        const val MAX_VIDEO_SIZE_BYTES = 50L * 1024 * 1024

        /** 128 bits of entropy per token. */
        private const val TOKEN_BYTES = 16

        /**
         * Opaque, unguessable capture token from the platform CSPRNG, hex-encoded.
         * The prior `"${millis}_${counter}"` scheme was enumerable — combined with
         * the WebView asset loader that would let crafted web content brute-force
         * other captures' URLs.
         */
        internal fun generateToken(): String =
            secureRandomBytes(TOKEN_BYTES).joinToString("") {
                ((it.toInt() and 0xFF) + 0x100).toString(16).substring(1)
            }
    }
}

/** Platform-agnostic epoch millis. */
internal expect fun currentTimeMillis(): Long

/** Fills and returns [count] bytes from the platform cryptographic RNG. */
internal expect fun secureRandomBytes(count: Int): ByteArray

File: lib/services/webview/capture_registry.dart
import 'dart:async';
import 'dart:io';

import 'package:snabbit_runner/services/monitoring/monitoring_service_helper.dart';

/// In-memory registry of captured images, keyed by opaque tokens with
/// automatic TTL-based expiry.
///
/// Captures are registered after the camera writes a file to the
/// dedicated temp directory, and resolved by [CapturePathHandler] when
/// the WebView fetches the image URL. Tokens expire after [_defaultTtl]
/// and a periodic sweep deletes stale files from disk.
///
/// Thread-safety: Dart is single-threaded (single isolate); no mutex
/// needed. The sweep timer fires on the event loop between awaits.
class CaptureRegistry {
  CaptureRegistry._();

  static final CaptureRegistry instance = CaptureRegistry._();

  static const Duration _defaultTtl = Duration(minutes: 15);
  static const Duration _sweepInterval = Duration(minutes: 5);

  /// Max servable capture size (10 MB). Single source of truth shared by
  /// the capture-time guard ([CaptureImageHandler]) and the serve-time
  /// guard ([CapturePathHandler]) so the two limits can't drift apart.
  /// A front-camera selfie is typically 1–3 MB; beyond this we can't load
  /// it into a [WebResourceResponse] without risking OOM on budget devices.
  static const int maxCaptureSizeBytes = 10 * 1024 * 1024;

  final Map<String, _CaptureEntry> _entries = {};
  Timer? _sweepTimer;

  /// Starts the periodic sweep timer if it isn't already running.
  void _ensureSweepTimer() {
    _sweepTimer ??= Timer.periodic(_sweepInterval, (_) => _sweep());
  }

  /// Cancels the sweep timer when there are no entries left to manage.
  void _cancelSweepTimerIfIdle() {
    if (_entries.isEmpty) {
      _sweepTimer?.cancel();
      _sweepTimer = null;
    }
  }

  /// Registers [file] and returns an opaque token the WebView can use to
  /// fetch the image via the asset-loader URL.
  ///
  /// If [token] is provided it is used as-is; otherwise one is generated
  /// from the current microsecond timestamp.
  String register(
    File file, {
    String? token,
    Duration ttl = _defaultTtl,
  }) {
    final key = token ?? DateTime.now().microsecondsSinceEpoch.toString();
    _entries[key] = _CaptureEntry(file, DateTime.now().add(ttl));
    _ensureSweepTimer();
    return key;
  }

  /// Returns the [File] for [token] if it is still valid, or `null` if
  /// the token is unknown or has expired. Expired entries are evicted
  /// eagerly (file deleted, entry removed).
  File? resolve(String token) {
    final entry = _entries[token];
    if (entry == null) return null;
    if (DateTime.now().isAfter(entry.expiresAt)) {
      _entries.remove(token);
      _deleteFile(entry.file);
      return null;
    }
    return entry.file;
  }

  /// Explicitly evicts [token] — deletes the backing file and removes
  /// the entry. No-op if the token is unknown.
  Future<void> evict(String token) async {
    final entry = _entries.remove(token);
    if (entry != null) {
      await _deleteFile(entry.file);
    }
    _cancelSweepTimerIfIdle();
  }

  /// Clears the entire capture [directory]. Intended to run **once at
  /// startup**: on a fresh process the in-memory registry is empty and no
  /// capture is in flight, so every file in `captures/` is an orphan from a
  /// prior session (a process kill can leave files the in-session TTL never
  /// evicted). Clearing wholesale is simpler and more complete than an age
  /// scan — no orphan lingers — and avoids stat-ing every file. The directory
  /// is recreated lazily on the next capture.
  ///
  /// Must NOT be called mid-session: it would delete files backing live
  /// registry entries.
  static Future<void> sweepStale(Directory directory) async {
    if (await directory.exists()) {
      await directory.delete(recursive: true);
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  void _sweep() {
    final now = DateTime.now();
    final expired = _entries.entries
        .where((e) => now.isAfter(e.value.expiresAt))
        .map((e) => e.key)
        .toList();

    for (final key in expired) {
      final entry = _entries.remove(key);
      if (entry != null) _deleteFile(entry.file);
    }

    if (expired.isNotEmpty) {
      MonitoringServiceHelper.logInfo(
        'capture_registry_sweep',
        {
          'evictedCount': expired.length,
          'remainingCount': _entries.length,
          'timestamp': DateTime.now().toIso8601String(),
        },
      );
    }

    _cancelSweepTimerIfIdle();
  }

  static Future<void> _deleteFile(File file) async {
    try {
      if (await file.exists()) await file.delete();
    } catch (_) {
      // Best-effort cleanup — the file may already be gone.
    }
  }
}

class _CaptureEntry {
  _CaptureEntry(this.file, this.expiresAt);

  final File file;
  final DateTime expiresAt;
}
